library/core/src/num/dec2flt/decimal.rs

```diff
@@ -0,0 +1,384 @@
+//! Arbitrary-precision decimal class for fallback algorithms.
+//!
+//! This is only used if the fast-path (native floats) and
+//! the Eisel-Lemire algorithm are unable to unambiguously
+//! determine the float.
+//!
+//! The technique used is "Simple Decimal Conversion", developed
+//! by Nigel Tao and Ken Thompson. A detailed description of the
+//! algorithm can be found in "ParseNumberF64 by Simple Decimal Conversion",
+//! available online: <https://nigeltao.github.io/blog/2020/parse-number-f64-simple.html>.
+
+use crate::fmt::{self, Debug};
+use crate::num::dec2flt::common::{is_8digits, parse_digits, ByteSlice, ByteSliceMut};
+
+#[derive(Clone)]
+pub struct Decimal {
+    /// The number of significant digits in the decimal.
+    pub num_digits: usize,
+    /// The offset of the decimal point in the significant digits.
+    pub decimal_point: i32,
+    /// If the sign of the float is negative.
+    pub negative: bool,
+    /// If the number of significant digits stored in the decimal is truncated.
+    pub truncated: bool,
+    /// Buffer of the raw digits, in the range [0, 9].
+    pub digits: [u8; Self::MAX_DIGITS],
+}
+
+impl Debug for Decimal {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Decimal")
+            .field("num_digits", &self.num_digits)
+            .field("decimal_point", &self.decimal_point)
+            .field("negative", &self.negative)
+            .field("truncated", &self.truncated)
+            .field("digits", &(&self.digits[..self.num_digits]))
+            .finish()
+    }
+}
+
+impl PartialEq for Decimal {
+    fn eq(&self, other: &Self) -> bool {
+        self.num_digits == other.num_digits
+            && self.decimal_point == other.decimal_point
+            && self.negative == other.negative
+            && self.truncated == other.truncated
+            && self.digits[..] == other.digits[..]
```

Comment:
The digits after are all zeroed out, so this correct, although likely sub-optimal. This likely hasn't been optimized since it's on a very slow code-path. However, this can and should be improved. 👍

See [here](https://github.com/Alexhuszagh/rust/blob/6bd33172108a39df2e6cc28fa248018f98d488b8/library/core/src/num/dec2flt/decimal.rs#L301-L304):

```rust
    for i in d.num_digits..Decimal::MAX_DIGITS_WITHOUT_OVERFLOW {
        d.digits[i] = 0;
    }
    d
```

Reply:
Actually, after carefully looking at the code, this is dead code: it's never used, and never exposed publicly. I'm just going to delete it.

Reply:
Actually, this has different logic. We can't do `self.num_digits <= Self::MAX_DIGITS`, because that would cause an indexing error in the cause where `self.num_digits == Self::MAX_DIGITS`, and we can't do `self.num_digits < Self::MAX_DIGITS`, since that would prevent incrementing to `MAX_DIGITS`, which then causes significant issues elsewhere.

A quick test on values with a lot of digits will demonstrate this:

```rust
let value = dec2flt::<f32>(format!("{}e-38", append_zeros(f1, 656))).unwrap();
assert_eq(value.to_bits(), 0x7ffffe);   // Fails, actual value is 0xccccd
```